tests/perf-system/analyzer/analyzer.py

```diff
@@ -0,0 +1,290 @@
+        )
+        return latency_output_table
+
+    def customize_table(self, fields_list: List[str], values_list: List[List]):
+        custom_table = PrettyTable()
+        custom_table.field_names = fields_list
+        for val_row in values_list:
+            custom_table.add_row(val_row)
+        return custom_table
+
+    def plot_latency_by_id(self, df_sends: pd.DataFrame) -> None:
+        id_unit = [x for x in range(0, len(df_sends.index))]
+        lat_unit = self.ms_latency_list
+        plt.figure()
+        plt.scatter(id_unit, lat_unit, s=1)
+        plt.ylabel("Latency_ms")
+        plt.xlabel("ids")
+        plt.savefig("latency_per_id.png")
+        plt.figure(figsize=(15, 15), dpi=80)
+
+    def plot_latency_across_time(self, df_responses) -> None:
+        time_unit = [
+            x - df_responses["receiveTime"][0] + 1 for x in df_responses["receiveTime"]
+        ]
+        plt.figure()
+        plt.scatter(time_unit, self.ms_latency_list, s=1)
+        plt.ylabel("Latency(ms)")
+        plt.xlabel("time(s)")
+        plt.savefig("latency_across_time.png")
+        plt.figure(figsize=(15, 15), dpi=80)
+
+    def plot_throughput_per_block(
+        self, df_responses: pd.DataFrame, time_block: float
+    ) -> None:
+        """
+        It splits the dataset in buckets of time_block seconds difference
+        and will plot the throughput for each bucket
+        """
+        # sort the latencies as it makes sense to get the throughput
+
+        # by time unit ignoring the ids
+        sorted_latencies = sorted(df_responses["receiveTime"].tolist())
+        block_indexes = [0]
+        for i, lat in enumerate(sorted_latencies):
+            if lat > sorted_latencies[block_indexes[-1]] + time_block:
+                block_indexes.append(i)
+        req_per_block = []
+        block_latency = []
+        if len(block_indexes) > 1:
+            for i in range(len(block_indexes) - 1):
+                req_per_block.append(block_indexes[i + 1] - block_indexes[i])
+                block_latency.append(time_block * SEC_MS * (i + 1))
+            req_per_block.append(len(sorted_latencies) - 1 - block_indexes[-1])
+            block_latency.append(
+                block_latency[-1]
+                + int(
+                    (sorted_latencies[-1] - sorted_latencies[block_indexes[-1]])
+                    * SEC_MS
+                )
+            )
+        throughput_per_block = [
+            x / time_block for x in req_per_block
+        ]  # x/time_block comes from rule of three
+        plt.figure()
+        plt.plot(block_latency, throughput_per_block)
+        plt.ylabel("Throughput(req/s)")
+        plt.xlabel("time(ms)")
+        plt.savefig("throughput_across_time.png")
+
+    def plot_latency_distribution(self, ms_separator: float, highest_vals=15):
+        """
+        Starting from minimum latency with ms_separator
+        step split the ms latency list in buckets
+        and plots the highest_vals top buckets

```

Comment:
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.plot.hist.html does this reasonably well?

Reply:
The problem I had with `pd.hist()` and other built in histogram functions is: 
1) It was complicated to remove 0 element bars
2) I wanted the bars to split in the step that the user would give e.g. every 0.1s and have bars [9.0, 9.1), [9.1, 9.2) e.t.c. Whereas with the built-in functions you can only specify how many bars/hists you want. That means that even after finding out how many bars you would have with the specific "step" (e.g. for a min_latency=9.067 and max_latency=10 with step 0.1 you would have 10 bars), the library will start the first bar with the first latency so the bars will be [9.067. 9.167), [9.167, 9.267) and so on, which I didn't want.

If that doesn't make sense and you want me to use pandas built-in please let me know.

Reply:
@fotiskoun no that's fine. I think you would do that by rounding your latencies in a separate column, counting the uniques and setting the number of bins to that, but if you have something that works, that's fine too.